axelrod/tests/unit/test_strategy_transformers.py

```diff
@@ -0,0 +1,184 @@
+        self.assertEqual(p2.history, [D, C, D, C, D])
+
+        p1 = axelrod.Alternator()
+        p2 = FlipTransformer(axelrod.Alternator)()
+        for _ in range(5):
+            p1.play(p2)
+        self.assertEqual(p1.history, [C, D, C, D, C])
+        self.assertEqual(p2.history, [D, D, D, D, D])
+
+
+    def test_initial_transformer(self):
+        # Initial play transformer
+        p1 = axelrod.Cooperator()
+        p2 = InitialTransformer([D, D])(axelrod.Cooperator)()
+        for _ in range(5):
+            p1.play(p2)
+        self.assertEqual(p2.history, [D, D, C, C, C])
+
+        p1 = axelrod.Cooperator()
+        p2 = InitialTransformer([D, D, C, D])(axelrod.Cooperator)()
+        for _ in range(5):
+            p1.play(p2)
+        self.assertEqual(p2.history, [D, D, C, D, C])
+
+    def test_final_transformer(self):
+        # Final play transformer
+        p1 = axelrod.Cooperator()
+        p2 = FinalTransformer([D, D, D])(axelrod.Cooperator)()
+        p2.tournament_attributes["length"] = 6
+        for _ in range(6):
+            p1.play(p2)
+        self.assertEqual(p2.history, [C, C, C, D, D, D])
+
+    def test_final_transformer2(self):
+        # Final play transformer (no tournament length)
+        p1 = axelrod.Cooperator()
+        p2 = FinalTransformer()(axelrod.Cooperator)()
+        for _ in range(6):
+            p1.play(p2)
+        self.assertEqual(p2.history, [C, C, C, C, C, C])
+
+    def test_history_track(self):
+        p1 = axelrod.Cooperator()
+        p2 = TrackHistoryTransformer(axelrod.Random)()
+        for _ in range(6):
+            p1.play(p2)
+        self.assertEqual(p2.history, p2._recorded_history)
+
+    def test_composition(self):
+        cls1 = InitialTransformer()(axelrod.Cooperator)
+        cls2 = FinalTransformer()(cls1)
+        p1 = cls2()
+        p2 = axelrod.Cooperator()
+        p1.tournament_attributes["length"] = 8
+        for _ in range(8):
+            p1.play(p2)
+        self.assertEqual(p1.history, [D, D, D, C, C, D, D, D])
+
+        cls1 = FinalTransformer()(InitialTransformer()(axelrod.Cooperator))
+        p1 = cls1()
+        p2 = axelrod.Cooperator()
+        p1.tournament_attributes["length"] = 8
+        for _ in range(8):
+            p1.play(p2)
+        self.assertEqual(p1.history, [D, D, D, C, C, D, D, D])
+
+    def test_retailiation(self):
+        RUA = RetailiateUntilApologyTransformer()
+        TFT = RUA(axelrod.Cooperator)
+        p1 = TFT()
+        p2 = axelrod.Cooperator()
+        p1.play(p2)
+        p1.play(p2)
+        self.assertEqual(p1.history, [C, C])
+
+        p1 = TFT()
+        p2 = axelrod.Defector()
+        p1.play(p2)
+        p1.play(p2)
+        self.assertEqual(p1.history, [C, D])
+
+        random.seed(12)
+        p1 = TFT()
+        p2 = axelrod.Random()
+        for _ in range(5):
+            p1.play(p2)
+        self.assertEqual(p1.history, [C, C, D, D, C])
+
+
+
+## Test that RUA(Cooperator) is the same as TitForTat
```

Comment:
This is the "proof" that the transformed class passes all the TFT tests by reusing the `TestTitForTat` test class from `test_titfortat.py`. However it fails the memory_depth test (since the transformer doesn't modify it, though perhaps it could). If I set the memory_depth manually, it actually modifies the classifier of the originally transformed class, since the classifier dict is a class variable until `__init__` is called, and is inherited by the transformed class. In this case I think the issue is that we are abusing that fact that classifiers are generally instance specific but in some cases identical for all instances (like for `Cooperator`), using a class variable when we should be using an instance variable.

The test code is fine once we decide what, if anything, to do about this -- it's not necessary to run the test for coverage, but it would be really nice to be able to include more proofs that many strategies can be decomposed into a series of transformations. So I left it in as a commented-out example and will add an issue.

One solution would be to add a method `register_classifiers` to the to the `Player` class. We're already modifying classifiers in `Player.__init__` and some strategies are modifying these further based on their initializing parameters. It's probably cleaner and better to do this with a single function, and to call this function at the end of `__init__`.